cpp/src/arrow/type.h

```diff
@@ -29,6 +29,7 @@
 
 #include "arrow/result.h"
 #include "arrow/type_fwd.h"  // IWYU pragma: export
+#include "arrow/util/bit_util.h"
```

Comment:
Is there a way to avoid including `bit_util.h` (which is a non-trivial header) here?

Reply:
Just do you want not to include `bit_util.h` in `type.h`. There are possible two ways.

One way is to include one of the following header files that explicitly include `bit_util.h`.
- buffer_builder.h
- type_traits.h
- visitor_inline.h

Or, try to move the definition of `class Endianness`. On the other hand, the header file that will newly define `class Endianness` requires including `bit_util.h`.

Reply:
If we just need endian-related definitions (such as `ARROW_LITTLE_ENDIAN` and `ARROW_BIG_ENDIAN`), then they can be moved to a separate `util/endian.h`, IMHO.

Reply:
I see. Should I create a separate PR for refactoring `util/endian.h`?

Reply:
No, it's minor enough that it can be done here.

Reply:
Newly added `util/endian.h` by moving some definitions from `bit_util.h`.